ouroboros-consensus/src/Ouroboros/Consensus/Node.hs

```diff
@@ -320,16 +321,20 @@ initNetwork registry nodeArgs kernel RunNetworkArgs{..} = do
                          then Just (Socket.SockAddrInet6 0 0 (0, 0, 0, 0) 0)
                          else Nothing
 
-    ipSubscriptions <- forkLinkedThread registry $
+    _ <- forkLinkedThread registry $
                          runIpSubscriptionWorker networkState ipv4Address ipv6Address
 
     -- dns subscription managers
-    dnsSubscriptions <- forM rnaDnsProducers $ \dnsProducer -> do
+    _ <- forM rnaDnsProducers $ \dnsProducer -> do
        forkLinkedThread registry $
          runDnsSubscriptionWorker networkState ipv4Address ipv6Address dnsProducer
 
-    let threads = localServer : ipSubscriptions : cleanPeerStatesThread : dnsSubscriptions ++ peerServers
-    void $ waitAnyThread threads
+    -- the only thread that terminates is 'cleanNetworkStateThread', all the
+    -- other threads return `Void`, thus they can only throw, since all of them
+    -- are linked to this thread we can just wait on the 'cleanNetworkStateThread'
+    -- Also note that 'cleanNetworkStateThread' terminates only if one of the
+    -- other threads throws an exception.
+    void $ waitThread cleanNetworkStateThread
```

Comment:
@mrBliss are you happy with this?

Reply:
That sounds ok. It would be even better if this logic would be somewhere in `Ouroboros.Network` instead of in `Ouroboros.Consensus`.

In fact, I think that `Ouroboros.Consensus.Node` still contains waaaay too much low-level networking setup (I moved it there from `cardano-node`) that should live in `Ouroboros.Network`.

Ideally, there would just be a single function that I need to call to start a server/client, instead of having to manually start multiple worker threads. They should all be internal to `Ouroboros.Network`. Changes in the network stack should not affect `Ouroboros.Consensus` at all.

I think `RunNetworkArgs` and the network-related functions in this file should move to `Ouroboros.Network`.

Reply:
This PR is already quite big, I'll address this in a follow up PR.